### DISCUSSION TITLE
Disconnected cameras & downloading events

### DISCUSSION THREAD

**Discussion by Sticklyman1936** [danielfernau/unifi-protect-video-downloader#62] [Q&A] [answered]
Hi there.

First of all, thanks a lot for creating this project. It works really well, and is very useful. I'm using it to sync recent events to cloud storage (in the event that someone walks off with my cloudkey I at least want a chance of getting that last bit of footage).

I was wondering if you could explain the logic behind only getting the `CONNECTED` cameras when getting the camera list. In the event that a camera goes offline but has events in the specified time period we get a StopIteration exception:

```
Getting motion event list
Successfully retrieved data from https://192.168.1.254:443/proxy/protect/api/events?type=motion&start=1629936000000&end=1630108800000
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/protect_archiver/downloader/get_motion_event_list.py", line 46, in <genexpr>
    "\n".join(f"{event_count_by_camera[x]} motion "
StopIteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/protect_archiver/cli/events.py", line 183, in events
    motion_event_list = client.get_motion_event_list(start, end, camera_list)
  File "/usr/local/lib/python3.8/site-packages/protect_archiver/client/__init__.py", line 79, in get_motion_event_list
    return Downloader.get_motion_event_list(self.session, start, end, camera_list)
  File "/usr/local/lib/python3.8/site-packages/protect_archiver/downloader/__init__.py", line 33, in get_motion_event_list
    return get_motion_event_list(session, start, end, camera_list)
  File "/usr/local/lib/python3.8/site-packages/protect_archiver/downloader/get_motion_event_list.py", line 46, in get_motion_event_list
    "\n".join(f"{event_count_by_camera[x]} motion "
RuntimeError: generator raised StopIteration
```
The end result is that one cannot download events for an offline camera (something which one might well want to do if replacing a camera). Worse still, one offline camera stops the download of any events from other cameras if the offline camera has an event in the start-end time period.

I've modified your code to include cameras irrespective of the state, and with the modification event download works as expected when one or more cameras are disconnected.

What have I broken by changing this logic? :)

Sascha

**Comment by danielfernau** [comment]
Hi Sascha –

Thank you for asking this, and I'm happy to hear that you like the software so far!
Yes, what you've encountered there is indeed a known issue and is (was?) caused by the way the Protect API returns the camera list. I haven't had a chance since then to check if there's been an update so that the API now returns the list of cameras correctly, including the ones that are offline.

Theoretically, there shouldn't be anything that you could have broken with your changes – I think. That part of the code has always been more of a workaround than a solution. Would you mind sharing your changes here (or opening a pull request) so that I can have a look at it?

Best
Daniel

**Reply by CodeChief** [reply]
Should be fixed in latest commit d6ba00111ab24789d6ce5ff3fc8d33dd2bb91506 of #89.

**Comment by danielfernau** [comment, marked as answer]
Hey everyone -- v2.1.3 is now available and includes the ability to download from disconnected cameras, thanks to @CodeChief's PR #89
